File: MMC Euclides.js
// 11/06/2019  - @Ericódigos
/*jshint esversion: 6 */ 

/**
  O exercício do FreeCodeCamp 'Intermediate Algorithm Scripting: Smallest Common 
  Multiple' solicitou-me uma função que retorne o Mínimo Múltiplo Comum de um 
  intervalo de inteiros. 
*/

/**
  Força bruta, a função delimita o maior multiplicador possível como o 
  produto de todos os elementos do array, de forma crescente e incrementando
  o número de vezes igual ao maior elemento do array. Cada possibilidade
  é testada em cada elemento do array retornando o menor múltiplica dor comum. 
*/
 const {performance} = require('perf_hooks'); //Para contar o tempo
 var iter; // Contador de iterações. 
 
 function f_MMC(x,y){  // Função Mínimo Multiplicador Comum
   // Classifica x e y e encontra o menor e o maior
   [x,y].sort( (a,b)=>{return a>b});
   // Cria 'arre' uma lista com todos os números inteiros entre X e Y inclusive.
   let arre = [];
   for(let i=arr[0];i<=arr[1];i++){
     arre.push(i); 
   }
   // Define 'pior' como o produto de todos elementos do array
   let pior = arre.reduce( (a,b)=>{return a*b});
   
/** 
  Inicia um loop a partir do maior elemento do conjunto (arre) até o limite
  definido por (pior), a cada iteração (j) é incrementado por sí mesmo e é 
  testado se o mesmo possui divisão de resto zero com todo os elementos de
  (arre), caso positivo retorne (j) que será o minímo múltiplo comum.
*/ 
   let v_lcm = false;
   iter = 0;
   for(let j=arre[arre.length-1];j<=pior;j+=arre[arre.length-1]){
     let v_lcm = true;
     iter++;
     for(let e in arre){
       iter++;
       if(j%arre[e]!==0){
         v_lcm = false
       }
     }
     if(v_lcm==true){
       return j;
     }
   }
 }
 var t0 = performance.now();
 console.log(f_MMC(23,15)) 
 var t1 = performance.now();
 console.log("A execução de 'f_MMC' durou " + (t1 - t0) + " milissegundos.");
 console.log("A execução de 'f_MMC' teve " + iter + " iterações.");


//     $$$$$$$$$$$$$$$$$$$$$$$    Exemplo okay?!  $$$$$$$$$$$$$$$$$$$$$$$$$$
var inter; // Contador de iterações.
function mmcIntervaloOkay(arr) { 
    // Classificando a lista do maior para o menor.
    inter=0;
    arr.sort(function(a, b) {
      return b - a;
    });
    // Criando uma lista com o intervalo dos inteiros entre o menor e o maior 
    // número fornecido.
    var newArr = [];
    for (var i = arr[0]; i >= arr[arr.length-1]; i--) {
      newArr.push(i);
    }
    // Variáveis fora do loop.
    var quot = 0; //Quociente 'Latin: Quantas Vezes'
    var loop = 1;
    var n;
    // Repita enquanto n for diferente do comprimento da lista.
    //let arre = [];
    do {
        inter++;
        quot = newArr[0] * loop * newArr[1]; //Produto dos dois maiores elementos da lista.
      //arre.push(quot)
      for (n = 2; n < newArr.length; n++) {
        inter++;
        if (quot % newArr[n] !== 0) {
          break;
        }
      }
  
      loop++;
    } while (n !== newArr.length);
    //console.log(arre)
    return quot;
  }
  
  // Chamada de teste
  var t0 = performance.now();
  console.log(mmcIntervaloOkay([23,18])); 
  var t1 = performance.now();
  console.log('A execução de mmcIntervaloOkay durou '+ (t1-t0) + ' milissegundos.');
  console.log("A execução de 'mmcIntervaloOkay' teve " + iter+ " iterações.");
/**
 *  11/06/2019
   Resultado: 6 056 820 
   A execução de mmcIntervaloOkay durou 4 milissegundos.
   A execução de 'mmcIntervaloOkay' teve 24570 iterações.
 */

//---------



// @#$@#$@#$@#$@#$ Matemática Euclidiana Suprema @#$@#$@#$@#$@#$@#$@#$
var iter=0; // Contador de iterações.

function mdc(a,b){
  iter++;
  if (b == 0){
      return a;
  }else{
      return mdc(b , a % b);
  }
}

function mmc(a,b){
  iter++;
  return ((a * b) / mdc(a,b));
}

function f_euclides_mmc(a,b){

  let arr = [a,b].sort( (a,b)=> a > b);
  let arre = [];
  for(let i=arr[0];i<=arr[1];i++){
    arre.push(i); 
  }
  console.log('O MMC do conjunto: [' + arre + '] é:');
  function f_mmc(cnj){
    iter++;
    if (cnj.length == 2){
        return mmc(cnj[0],cnj[1]);
    }else{
        ar = [mmc(cnj[0],cnj[1]),...cnj.slice(2)];
        return f_mmc(ar);
    }
  }
  return f_mmc(arre);
}

var t0 = performance.now();
console.log(f_euclides_mmc(23, 18));
var t1 = performance.now();
console.log('A execução de f_euclides_mmc durou '+ (t1-t0) + ' milissegundos.');
console.log("A execução de 'f_euclides_mmc' teve " + iter+ " iterações.");

//       ~~$#$%**~~~~$#$%**~~ Fim do exemplo supremo ~~$#$%**~~~~$#$%**~~ 

File: MMC Euclides.py
'''@Ericódigos - 11/06/2019 
Fundamentos são importantes. Uma prova disse é um algorítimo 
escrito a mais de 2300 anos por Euclides e que ainda hoje é 
absolutamente útil e elegantemente econômico em termos de 
processamento.

Imagine um problema que exija uma função que retorne o 
Mínimo Múltiplo Comum de um intervalo de inteiros.

'''

import functools
from datetime import datetime

time1 = datetime.now()
lista = []

def mmc_Rank_C(lst):
  """ Retorna MMC de um intervalo de inteiros. """
  itera = 0 # Contador de interações
  lst.sort() # Classifica a lista
  
  ''' Cria a lista com o intervalo entre o menor e o maior número. '''
  lista = list(range(lst[0],lst[-1]+1))
  
  ''' Retorna o fatorial do intervalo'''
  pior = functools.reduce(lambda a,b: a*b,lista)

  for e in range(0,pior,lista[-1]):
    itera += 1
    if all(e % i == 0 and e != 0 for i in lista):
      print('Iterações: ',itera)
      print('Resultado do MMC de ', lista) 
      return e
  

print('-->',mmc_Rank_C([23,15]))
time2 = datetime.now()
print('A função demorou: ',(time2-time1))



###########################



time1 = datetime.now()

def mmc_Rank_B(lista):
  lista.sort(reverse=True)
  novaLista = list(range(lista[0],lista[-1]-1,-1))
  quociente = 0
  loop = 1
  n = 0
  while(n != len(novaLista)):
    quociente = novaLista[0] * loop * novaLista[1]
    if all(quociente % e == 0 for e in novaLista):
      break
    loop+=1
  print('Iterações: ',loop)  
  return quociente
  
print(mmcIntervaloOkay([23,15]))
time2 = datetime.now()
print('A função demorou: ',(time2-time1))


###################################


time1 = datetime.now()
itera = 0
def mmcIntervalo_Rank_S(lst):
  global itera
  mínimo = min(lst)
  máximo = max(lst)
  v_mmc = mmc(mínimo, mínimo+1)
  lista = list(range(mínimo,máximo))
  while(mínimo < máximo):
    mínimo += 1
    v_mmc = mmc(v_mmc,mínimo)
  print('Iterações: ',itera)
  print('O MMC de: ',lista)
  return v_mmc  

def mdc(a,b):
  global itera
  while(b != 0):
    itera += 1
    t = a
    a = b
    b = t % b
  return a

def mmc(a,b):
  return int((a * b /mdc(a,b)))

print(mmcIntervalo_Rank_S([23,15]))
time2 = datetime.now()
print('A função demorou: ',(time2-time1))














''' math.gcd(a, b) '''
